ProjFS.Mac/PrjFSLib/PrjFSLib.cpp

```diff
@@ -1031,3 +1056,34 @@ static const char* NotificationTypeToString(PrjFS_NotificationType notificationT
     }
 }
 #endif
+
+static shared_ptr<mutex> CheckoutPathMutex(const string& fullPath)
+{
+    mutex_lock lock(s_inProgressExpansionsMutex);
+    PathToMutexMap::iterator iter = s_inProgressExpansions.find(fullPath);
+    if (iter == s_inProgressExpansions.end())
+    {
+        pair<PathToMutexMap::iterator, bool> newEntry = s_inProgressExpansions.insert(
+            PathToMutexMap::value_type(fullPath, { make_shared<mutex>(), 1 }));
+        assert(newEntry.second);
+        return newEntry.first->second.mutex;
+    }
+    else
+    {
+        iter->second.useCount++;
+        return iter->second.mutex;
+    }
+}
+
+static void ReturnPathMutex(const string& fullPath, const shared_ptr<mutex>& mutex)
```

Comment:
It's a fairly minor point, but given that `CheckoutPathMutex`/`ReturnPathMutex` are always paired in a function, I'm wondering if `CheckoutPathMutex` should return an iterator/mutex `pair`, and `ReturnPathMutex` would accept that iterator to avoid looking up the fullPath afresh.

Reply:
That's a good idea thanks!